keps/sig-apps/2804-consolidate-workload-controllers-status/README.md

```diff
@@ -0,0 +1,376 @@
+Since a DaemonSet can make progress no faster than "healthy but not ready nodes", the default value for `progressDeadlineSeconds` can be set to 30 minutes but this value can vary depending on the node count in the cluster.
+The value for StatefulSet can be longer than 10 minutes since it also involves provisioning storage and binding. This default value can be set to 15 minutes in case of StatefulSet.
+
+It is possible that we introduce a bug in the implementation. The bug can cause:
+- DaemonSet and StatefulSet controllers can be marked `Failed` even though rollout is in progress
+- The states could be misrepresented, for example a DaemonSet or StatefulSet can be marked `Progressing` when actually it is complete
+
+The mitigation currently is that these features will be in Alpha stage behind `ExtendedWorkloadConditions` featuregate for people to try out and give feedback. In Beta phase when
+these are enabled by default, people will only see issues or bugs when `progressDeadlineSeconds` is set to something greater than 0 and we choose default values for that field.
+Since people would have tried this feature in Alpha, we would have had time to fix issues.
+
+
+## Design Details
+
+### Test Plan
+Unit and E2E tests will be added to cover the
+API validation, behavioral change of controllers with feature gate enabled and disabled.
+
+- Validating all possible states of old and new conditions. Checking that the changes in underlying Pod statuses correspond to the conditions.
+- Testing `progressDeadlineSeconds` and feature gates.
+
+
+### Graduation Criteria
+
+#### Alpha
+- Complete feature behind featuregates
+- Have proper unit, integration and e2e tests 
+
+#### Alpha -> Beta Graduation
+- Gather feedback from end users
+- Tests are in Testgrid and linked in KEP
+- all new features in the following controllers should be implemented: ReplicaSet & ReplicationController, StatefulSet, DaemonSet and Job. To fully support `failed` state of a progressing condition, `progressDeadlineSeconds` KEP should be also fully implemented.
+
+#### Beta -> GA Graduation
+- 2 examples of end users using this field
+
+### Upgrade / Downgrade Strategy
+
+- Upgrades
+ When upgrading from a release without this feature, to a release with `ExtendedWorkloadConditions` feature,
+ we will set new conditions on the mentioned workloads.
+- Downgrades
+ When downgrading from a release with this feature, to a release without, 
+ we expect controllers not to remove stale conditions.
+
+### Version Skew Strategy
+
+The update of extended conditions is always dependent on a `ExtendedWorkloadConditions` feature gate and not on the version as such.
+If the feature gate is enabled, the workload controllers will update the extended conditions to reflect the current state.
+In case feature gate is disabled, the workload controllers will remove the stale conditions.
+In case the feature is missing in the workload controllers the conditions will not be removed and become stale.
```

Comment:
I would expect the behaviour to be the same if either the feature is not implemented (old version) or the feature gate is disabled. Otherwise, we will have changed default behaviour.

Can you confirm if this is the case? Are stale conditions already removed without this feature, or is this a net new change?

Reply:
@ehashman Unfortunately we checked the controllers and most of them currently don't do updates of the conditions in the status. So it is not easy to ensure their removal.

Reply:
I think we don't want removal, as that would change default behaviour.